src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/Source.cs

```diff
@@ -0,0 +1,238 @@
+                            ? description
+                            : string.Empty))
+                    : null;
+
+            // This is a code supporting legacy completion scenarios:
+            // If there are suggestionItemOptions, then later HandleNormalFiltering should set selection to SoftSelection.
+            session.Properties.AddProperty(HasSuggestionItemOptions, suggestionItemOptions != null);
+
+            return new AsyncCompletionData.CompletionContext(
+                items,
+                suggestionItemOptions,
+                suggestionItemOptions == null ? AsyncCompletionData.InitialSelectionHint.RegularSelection : AsyncCompletionData.InitialSelectionHint.SoftSelection);
+        }
+
+        public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, VSCompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextSnapshot>(TriggerSnapshot, out var triggerSnapshot))
+            {
+                return null;
+            }
+
+            var document = triggerSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var service = document.GetLanguageService<CompletionService>();
+            var description = await service.GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
+
+            return new ClassifiedTextElement(description.TaggedParts.Select(p => new ClassifiedTextRun(p.Tag.ToClassificationTypeName(), p.Text)));
+        }
+
+        private static RoslynTrigger GetRoslynTrigger(AsyncCompletionData.CompletionTrigger trigger, SnapshotPoint triggerLocation)
+        {
+            switch (trigger.Reason)
+            {
+                case AsyncCompletionData.CompletionTriggerReason.Invoke:
+                case AsyncCompletionData.CompletionTriggerReason.InvokeAndCommitIfUnique:
+                    return RoslynTrigger.Invoke;
+                case AsyncCompletionData.CompletionTriggerReason.Insertion:
+                    return RoslynTrigger.CreateInsertionTrigger(trigger.Character);
+                case AsyncCompletionData.CompletionTriggerReason.Deletion:
+                    return RoslynTrigger.CreateDeletionTrigger(trigger.Character);
+                case AsyncCompletionData.CompletionTriggerReason.Backspace:
+                    var snapshotBeforeEdit = trigger.ViewSnapshotBeforeTrigger;
+                    char characterRemoved;
+                    if (triggerLocation.Position >= 0 && triggerLocation.Position < snapshotBeforeEdit.Length)
+                    {
+                        // If multiple characters were removed (selection), this finds the first character from the left. 
+                        characterRemoved = snapshotBeforeEdit[triggerLocation.Position];
```

Comment:
is this the previous behavior?

Reply:
We haven't had such case in the old completion. There in case of backspace, we already had the character deleted.

Here, Editor provides us \b in case of backspace. However, we need the character just removed.

---
In reply to: [222522644](https://github.com/dotnet/roslyn/pull/29016#discussion_r222522644) [](ancestors = 222522644)